Grue changes: different moulting mechanics, exposure-stacking light damage, basic NPC AI, various other tweaks.

Hinaichigo opened the pull request:
In accordance with feedback from [this thread](http://ss13.moe/forum/viewtopic.php?f=4&t=6167), this refactors the grue light-based damage mechanics so that they take more damage per tick the longer they have recently been exposed to light, making brief sallies out into the light less punishing but continually increasing damage the longer the grue remains exposed.

Also adds some basic AI to NPC grues that prevents them from wandering into light, and lets them moult, eat, and reproduce.  The eggs will only hatch if they successfully recruit a player ghost though.

Also removes the shadow power mechanic and gives juveniles the ability to eat sentient beings as well, and successfully doing so makes them stronger and gives them the nutrients required to moult into adulthood. So they are more like weaker adults that can't lay eggs, and now aren't incentivized just to lurk in the dark until they are fully mature. Hopefully this makes them more interesting to play as.

Edit: I tagged this incorrectly
[content][tweak]

:cl:
 * rscadd: Grues now take initially lower but increasing damage the longer they spend exposed to light.
 * rscadd: Grues get a temporary healing boost after feeding.
 * rscadd: Juvenile grues can now eat like adults and must feed once in order to moult into an adult.  
 * rscadd: Added basic AI for NPC grues.
 * rscdel: Removed shadow power mechanic.
 * tweak: Grues heal faster in the dark.
 * tweak: Grues can pull other creatures and objects.
 * tweak: Grues can eat a little bit faster.

Eneocho commented:
>The eggs will only hatch if they successfully recruit a player ghost though.

I was kinda looking forward to NPC gruening, like the NPC spidering that happens after a player spider lays eggs.

Hinaichigo commented:
> > The eggs will only hatch if they successfully recruit a player ghost though.
> 
> I was kinda looking forward to NPC gruening, like the NPC spidering that happens after a player spider lays eggs.

I considered it as well but cycling dead players back into the round seems more fun, maybe.

Eneocho commented:
make it spawn as simplemob if there was a mobcall but nobody joined.

Eneocho commented:
Also, can you increase the door opening speed? I observed a grue and it took way too long for the guy to open doors as a new adult. 
Apparently it gets faster once you eat 5 players, but by then you don't need to open doors, you can crush walls.

WindowsErrors commented:
out of all the 10,000 human languages that eneocho could have spoken they decided to speak facts

if you think about it in terms of other reproducing maintenance-prowler antagonists like xenos or spiders, grues are definitely the slowest since they have the least benefits that come to them already being pre-established when you spawn in (like what, you get a deeper ass crack of maintenance to grow for 10 minutes in?) as opposed to xenos (the increased plasma output of the hive you spawn in means the only real waiting being done is by the original queen who set your place up and waited for you to sign up) or spiders (who are already going to spawn in areas spread out across the station, leaving very little downtime for you or the crew when more than one of them gets swole and in control)

Hinaichigo commented:
> Can you increase the door opening speed? I observed a grue and it took way too long for the guy to open doors as a new adult. Apparently it gets faster once you eat 5 players, but by then you don't need to open doors, you can crush walls.

Currently it takes a juvenile 10 seconds and a yet-to-eat adult 5 seconds, removing 2 seconds every time the adult eats, so it becomes instant after 3 eatings. I think reducing it to 5 for juveniles and 3 for adults (-1 second each feeding) could be a good balance, so I went ahead and did that. I do want to avoid it being instant though so they have to be a bit more cautious about venturing into the light.

As for the waiting time, the juvenile state is sort of meant to be a hybrid state between the cautious, hiding mode of the larva and the more aggressive, hungry state of the adult. Not necessarily just for waiting around, though I guess people are incentivized to just hide and not make their presence known until they reach their most powerful form.
One solution to this I was thinking of was giving the juvenile the Eat ability as well and they would instantly morph into an adult upon eating a sentient, (or maybe be able to moult much faster than normal) so there would be some incentive for riskier plays and not just hiding to fully power up before being aggressive.

Some numbers for the time to reach adult:
In full darkness a grue gains 3 shadow power per tick, so it takes:
~83 seconds (1:23) to get the 250 shadow power necessary to moult into a juvenile.
30 seconds to moult
 ~167 (2:47) seconds to get the 500 shadow power necessary to moult into an adult
30 seconds to moult.
So it takes 5 minutes and 10 seconds to go from larva to adult.

Let me know what kind of adjustments you'd like to see on those times.

Eneocho commented:
You could add some minor buffs if the grue fed as a teen (better light endurance? eating lights for a sec?), so there's an incentive.
Prime Grues are those that fed as teens and have the extra stuff. Regular grues are those that just became adults by waiting.

At adult, the guy I observed couldn't heal the damage in time, it ticked too slow and since he couldn't eat a single sentient, he also didn't get any healing (maybe add healing via eating monkeys?). 
You could reuse the shadowpower from baby and teen as a ability "fuel" for the adult stage. You actively accumulate shadowpower as adult too and can use it to power some more advanced shit, like the Umbra (former grue) light eating power, enhanced healing, etc.

Eneocho commented:
Also, I'd slice the waiting times for baby in half, the teen one by a little too.

Hinaichigo commented:
Interesting ideas. One thing I almost added to this PR was a temporary healing boost after a successful feeding, but decided against it. It was such that after eating you would heal an extra 2 health per second for 25 seconds. What do you think about that? And what do you and others think about having there not be a sentient check for feeding in general. So if a grue breaks into virology or genetics during a power outage or something you have a catastrophe on your hands.

Overall I erred on the side of making them underpowered at first (the EI NATH thing notwithstanding), but I guess, indeed, people want them to be somewhat stronger?

Eneocho commented:
The removal of the sentient check would be interesting, it'd become an alternate spider (since spiders just need to feed on living beings, no matter the sentience). 
I'd make the feeding on non-sentients contribute less than feeding on sentients. Say, 2 monkeys = 1 sentient for point purposes, so you CAN get to the advanced levels, but you are encouraged to fight against crew for faster growth.
That way it'd be a viable if slow way to reach fuckgrue levels.
Do add the healingbuff, either via spell and accumulated points or via feeding on anything.

Hinaichigo commented:
What do you think about increasing brute resistance as grues power up?

Eneocho commented:
Not sure. You mean as they eat more and more? You could increase the resistance by just a bit whenever they feed, so the 5 sentients fuckgrue is a force to be reckoned with and not a casual fight (similar to queens in both beno and spiders).
Try it, if it doesn't work, you can just remove it later.

Hinaichigo commented:
Doubled the base healing rate and added 75 health worth of healing over 25 seconds per sentient feed. Also reduced the larva->juvenile shadow power-time from 83 to 50 seconds and juvenile to 100 seconds (1:40).
I hesitate to make it too fast though, otherwise people might not try to do anything as larva and juveniles and just wait until they can moult, type of thing.

WindowsErrors commented:
but isn't that what the point of a larva is? laying low until you're ready to strike out?

Hinaichigo commented:
Good point. Also the ventcrawling lets you scout the map and see where you might want to moult and base your operations in. So maybe 30 seconds for larvae and 1:00 or 1:30 for juveniles, or something like that?

Hinaichigo commented:
Changed it so that juvenile grues can also Eat sentients. It takes them ~2:30 to get enough shadowpower to moult but eating a sentient will give half of the necessary shadowpower over 25 seconds. So this should hopefully make juveniles more like weaker, non-egglaying adults and maybe more dynamic to play as.
Also removed the airlock forcing time across the board. I think factoring in players chasing them and stuff it's probably fine to have it instant.

Hinaichigo commented:
Okay, I changed the shadowpower mechanic into "nutritive energy." Basically a larva starts out already nourished from the egg yolk so they can moult whenever they want. It's up to the player if they want to moult where they spawn or if they want to scout the map for a more optimal location by ventcrawling and exploring.
And juveniles start out hungry can only moult after consuming one sentient being. They can't reproduce until doing so either. Now when the grue herd kills someone they might end up squabbling over who gets to eat the carcass, which could be interesting and adds to the hunger vibes; a juvenile could eat it to grow into an adult, or an adult could eat it to get even stronger.

Eneocho commented:
Very nice.
Here's hoping grue get some special powers eventually, but that's for another PR.

DamianX commented on code/modules/mob/living/simple_animal/grue_egg.dm:
```diff
@@ -35,6 +35,11 @@
 	var/ping_cooldown = 50
 
+	var/accum_light_expos_mult= 1 //used to scale light damage the longer the grue egg is exposed to light
+	var/list/accum_light_expos_gain_dark_dim_light=list(-3,-1,1) //light damage rate increases the longer the grue egg is exposed to light, but this effect dissipates after going back into darkness
```
This var and others are duplicated between `grue_egg` and `grue`.
Have you considered using inheritance to avoid repeating yourself?

Hinaichigo replied:
One issue is that grue eggs are  `mob/living/simple_animal` while grues are `mob/living/simple_animal/hostile`, so I'm not sure what the ideal solution is.

DamianX replied:
True. You could move these vars and procs to a datum and use it in both places.

Hinaichigo replied:
I think I got it set up as a datum.

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -137,13 +166,81 @@
 	standard_damage_overlay_updates()
 
+
+	if((stat!=DEAD) && (stat!=UNCONSCIOUS) && !busy && !ismoulting && !client && !mind && ckey==null) //Checks for AI
```
```suggestion
	if((stat == CONSCIOUS) && !busy && !ismoulting && !client && !mind && !ckey) //Checks for AI
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -137,13 +166,81 @@
+					if(ourdest)
+						wander_move(ourdest)
+						turns_since_move = 0
+						INVOKE_EVENT(src, /event/after_move)
```
`wander_move` will call `Move` which should invoke this event. Is this necessary for some reason?

Hinaichigo replied:
I copied it along from elsewhere, I'll remove it.

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -333,6 +439,27 @@
+
+/mob/living/simple_animal/hostile/grue/proc/nutri_adjust()
+	switch(dark_dim_light)
+		if(0) //if dark
```
These should use the newly-#defined macros

Hinaichigo commented:
Moved the light-related parameters to a datum shared by both grues and grue eggs.

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -21,45 +21,112 @@
+	var/busy=0 //busy attempting to lay an egg or eat
 
 	var/eattime= 5 SECONDS //how long it takes to eat someone
+	var/digest = 0 //how many seconds of healing left after feeding
```
This seems to be misleading. The value is not expressed in seconds but in number of Life() ticks, which currently happen every 2 seconds.

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -21,45 +21,112 @@
+	var/datum/grue_calc/grue/lightparams = new /datum/grue_calc/grue //used for light-related calculations
+
+	//AI related:
+	stop_automated_movement = 1 //has custom light-related wander movement
```
```suggestion
	stop_automated_movement = TRUE //has custom light-related wander movement
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -21,45 +21,112 @@
+
+	//AI related:
+	stop_automated_movement = 1 //has custom light-related wander movement
+	wander=0
```
```suggestion
	wander = FALSE
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -137,13 +210,80 @@
+			turns_since_move++
+			if(turns_since_move >= turns_per_move)
+				if(!(stop_automated_movement_when_pulled && pulledby))
+					INVOKE_EVENT(src, /event/before_move)
```
`wander_move` should call `Move` which should invoke this event, why is this here?

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -137,13 +210,80 @@
+					var/list/potential_dests=list() //any potential wander destinations
+					var/list/respite_dests=list() //any potential wander destinations leading out of the light into dark/dim
+					for(var/thiscardinal in 1 to cardinal.len)
+						var/potential_dest = get_step(src, cardinal[thiscardinal])
```
```suggestion
					for(var/direction in cardinal)
						var/potential_dest = get_step(src, direction)
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -137,13 +210,80 @@
+								potential_dests+=potential_dest
+					var/ourdest
+					if(respite_dests.len)
+						ourdest=(pick(respite_dests))
```
```suggestion
						ourdest=pick(respite_dests)
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -137,13 +210,80 @@
+					if(respite_dests.len)
+						ourdest=(pick(respite_dests))
+					else if(potential_dests.len)
+						ourdest=(pick(potential_dests))
```
```suggestion
						ourdest=pick(potential_dests)
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -21,45 +21,112 @@
-	var/dark_dim_light=0 //darkness level currently the grue is currently exposed to, 0=nice and dark, 1=passably dim, 2=too bright
-	var/busy=0 //busy laying an egg
+
+	var/busy=0 //busy attempting to lay an egg or eat
```
```suggestion
	var/busy=FALSE //busy attempting to lay an egg or eat
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -410,50 +556,63 @@
+		to_chat(src, "<span class='danger'>You have eaten \the [E]!</span>")
 		to_chat(E, "<span class='danger'>You have been eaten by a grue.</span>")
 
+		digest+=25 //add 25 seconds of increased healing + nutrition gain
```
See earlier comment about this var. This will add 50 seconds, not 25, I think.

Hinaichigo replied:
Oh, I had assumed Life() was 1 second.

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -410,50 +556,63 @@
+
+	if(lifestage==GRUE_ADULT)
+		if(eatencount>=GRUE_WALLBREAK)
+			environment_smash_flags = environment_smash_flags | SMASH_WALLS
```
```suggestion
			environment_smash_flags |= SMASH_WALLS
```

DamianX commented on code/modules/mob/living/simple_animal/hostile/grue.dm:
```diff
@@ -410,50 +556,63 @@
+		if(eatencount>=GRUE_WALLBREAK)
+			environment_smash_flags = environment_smash_flags | SMASH_WALLS
+		if(eatencount>=GRUE_RWALLBREAK)
+			environment_smash_flags = environment_smash_flags | SMASH_RWALLS
```
```suggestion
			environment_smash_flags |= SMASH_RWALLS
```

Eneocho commented:
Please make it so teen and adult grues can drag their victims. All other "monster" antags can.

Eneocho commented:
Could you eventually add that light sucking skill that humanoid grues (now umbra) had, so you can make a mad rush against someone with lights on them?

Hinaichigo commented:
> Could you eventually add that light sucking skill that humanoid grues (now umbra) had, so you can make a mad rush against someone with lights on them?

I've been thinking about it and whether and how to implement that... though if I can get #32007 working it might make it less necessary, especially if grues can target specific bodyparts. Then they could bite someone's arm to make them drop their lantern, then smash the lantern once it's on the ground. 
Also, though it might be interesting mechanically, I hesitate because I wonder if the dark aura would be flavor-congrueous with the current iteration of grues; it would sort of make them more shadow-elemental, ish, which is something I've been avoiding.

Eneocho commented:
I'd like to see it though, it would work mechanically and as an homage to the original grue (Umbra).
Try it, worst that can happen is it doesn't work and we revert it.

Hinaichigo commented:
Maybe we can come up with a bunch of different powers post-larval grues can use, and the shadow aura could be one of them. If we add that we could also add invisibility in full (not dim) darkness. But like you said it's for a different PR.
I added the pulling.
Also, should umbras be renamed to gruekin or something like that, do you think? Gruoids..?

Eneocho commented:
I think umbra is fine, though I use both for now since we're still at that stage where the two variants can be confused.

Hinaichigo commented:
> I think umbra is fine, though I use both for now since we're still at that stage where the two variants can be confused.

By the way if you want to, please post some ideas for abilities and mechanics in #31969.

DamianX reviewed (approved):
Is this ready?

Hinaichigo commented:
> Is this ready?

Yes, ready.
